File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章正则问题.py
#正则问题 2017年第八届省赛C/C++A组   
#lanqiaoOJ 106
s=input().strip()
pos,length=0,len(s)
def dfs():
    global pos,length
    ans,temp=0,0
    while pos<length:
        if   s[pos]=='(':   pos+=1; temp=temp+dfs()        
        elif s[pos]=='x':   pos+=1; temp+=1
        elif s[pos]=='|':   pos+=1; ans=max(ans,temp); temp=0
        elif s[pos]==')':   pos+=1; return max(ans,temp)
    return max(ans,temp)
print(dfs())

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章正则问题.java
//正则问题 2017年第八届省赛C/C++A组   
//lanqiaoOJ 106
import java.util.Scanner;
 public class Main {
    static String s;
    static int pos;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        s = sc.nextLine();   
        System.out.println(dfs());
    }
    public static int dfs() {
        int ans = 0;
        int temp = 0;
        int len = s.length();
        while (pos < len) {
            if(s.charAt(pos) == '(') {
                pos++;
                temp += dfs();
            } else if(s.charAt(pos) == 'x') {
                temp++;
                pos++;
            }else if(s.charAt(pos) == '|') {
                ans = Math.max(ans,temp);
                temp = 0;
                pos++;
            } else if(s.charAt(pos) == ')') {
                pos++;
                return Math.max(temp,ans);
            }
        }
        return Math.max(ans,temp);
    }
}
